viewer/packages/camera-manager/src/CameraManager.ts

```diff
@@ -0,0 +1,534 @@
+
+  public automaticNearFarPlane: boolean = true;
+  public automaticControlsSensitivity = false;
+
+  private isDisposed = false;
+  private readonly _animationDuration: number = 600;
+  private readonly _minDefaultAnimationDuration: number = 600;
+  private readonly _maxDefaultAnimationDuration: number = 2500;
+  private readonly _minDistanceDefault: number = 0.1;
+  private readonly _raycaster: THREE.Raycaster = new THREE.Raycaster();
+  private _onClick: ((event: MouseEvent) => void) | undefined = undefined;
+  private _onWheel: ((event: MouseEvent) => void) | undefined = undefined;
+
+  /**
+   * Reusable buffers used by functions in Cognite3dViewer to avoid allocations.
+   */
+  private readonly _updateNearAndFarPlaneBuffers = {
+    cameraPosition: new THREE.Vector3(),
+    cameraDirection: new THREE.Vector3(),
+    nearPlaneCoplanarPoint: new THREE.Vector3(),
+    nearPlane: new THREE.Plane(),
+    corners: new Array<THREE.Vector3>(
+      new THREE.Vector3(),
+      new THREE.Vector3(),
+      new THREE.Vector3(),
+      new THREE.Vector3(),
+      new THREE.Vector3(),
+      new THREE.Vector3(),
+      new THREE.Vector3(),
+      new THREE.Vector3()
+    )
+  };
+
+  /**
+   * Convert pixel coordinates of the cursor to [-1,1]^2 coordinates.
+   * @param pixelX
+   * @param pixelY
+   */
+  private convertPixelCoordinatesToNormalized(pixelX: number, pixelY: number) {
+    const x = (pixelX / this._domElement.clientWidth) * 2 - 1;
+    const y = (pixelY / this._domElement.clientHeight) * -2 + 1;
+
+    return { x, y };
+  }
+
+  /**
+   * Calculates new target when raycaster doesn't have any intersections with the model.
+   * @param cursorPosition Cursor position for desired calculations.
+   * @param cursorPosition.x
+   * @param cursorPosition.y
+   */
+  private calculateMissedRaycast(
+    cursorPosition: { x: number; y: number },
+    modelsBoundingBox: THREE.Box3
+  ): THREE.Vector3 {
+    const modelSize = modelsBoundingBox.min.distanceTo(modelsBoundingBox.max);
+
+    this._raycaster.setFromCamera(cursorPosition, this._camera);
+
+    const farPoint = this._raycaster.ray.direction
+      .clone()
+      .normalize()
+      .multiplyScalar(
+        Math.max(this._camera.position.distanceTo(modelsBoundingBox.getCenter(new THREE.Vector3())), modelSize)
+      )
+      .add(this._camera.position);
+
+    return farPoint;
+  }
+
+  /**
+   * Changes controls target based on current cursor position.
+   * @param event MouseEvent that contains pointer location data.
+   */
+  private async changeTarget(event: MouseEvent) {
+    const { offsetX, offsetY } = event;
+
+    const { x, y } = this.convertPixelCoordinatesToNormalized(offsetX, offsetY);
+
+    const callbackData = await this._modelRaycastCallback(offsetX, offsetY);
+
+    const newTarget =
+      callbackData?.intersection?.point ?? this.calculateMissedRaycast({ x, y }, callbackData.modelsBoundingBox);
```

Comment:
```suggestion
    const newTarget =
      callbackData.intersection?.point ?? this.calculateMissedRaycast({ x, y }, callbackData.modelsBoundingBox);
```
`callbackData` is never `undefined` (?)

Reply:
Should be, yeah

Reply:
Oh, it is undefined if raycast misses the model actually, so should be with question mark.

Reply:
Ok, missed it for some reason